ls1intum/Artemis
# Performance/limit plagiarism comparisons

Exercises with many students can lead to extreme performance problems, because millions of objects would need to be saved in the database. This does not make sense, because no now would view those plagiarism comparisons anyway. Also, it can have severe performance issues and even break the system completely. Therefore, this PR limits the saved and returned results to the 500 with the highest similarity.

### Checklist

### Steps for Testing
1. Create many submissions (>40) for a modeling, text or programming exercise
2. Invoked the plagiarism check on the exercise detail page with the minimal similarity setting (50).
3. Make sure only up to 500 results are shown and the whole operation does not take too much time.

(Note that we have up to 2.200 submissions per exercise in courses on production, which could potentially lead to 2.200 * 2.200 / 2 = 2.420.000 comparisons). While the comparisons themselves are rather fast, saving the values in the database is time-consuming.

@iozbeyli helped me to track some performance metrics with 40 similar realistic modeling submissions
* Comparing plagiarisms took 64ms and led to 728 comparisons. Saving those 728 plagiarism comparisons in the database took around 10,4 s
